File: mobile_client/lib/widgets/form_sheet.dart
import 'package:flutter/material.dart';

import '../theme.dart';

/// An OPAQUE modal bottom sheet for create/edit forms. Opaque
/// ([JcTheme.surface]) is deliberate — translucent sheets bleed the aurora
/// backdrop through (the recurring JARVIS-skin gotcha).
///
/// [onSave] returns true on success (sheet pops with `true`) or false to keep
/// the sheet open (e.g. validation failed and a SnackBar was shown).
Future<bool?> showFormSheet({
  required BuildContext context,
  required String title,
  required List<Widget> fields,
  required Future<bool> Function() onSave,
  String saveLabel = 'Save',
}) {
  return showModalBottomSheet<bool>(
    context: context,
    isScrollControlled: true,
    backgroundColor: JcTheme.surface,
    shape: const RoundedRectangleBorder(
      borderRadius: BorderRadius.vertical(top: Radius.circular(22)),
    ),
    builder: (ctx) => _FormSheetBody(
      title: title,
      fields: fields,
      onSave: onSave,
      saveLabel: saveLabel,
    ),
  );
}

class _FormSheetBody extends StatefulWidget {
  const _FormSheetBody({
    required this.title,
    required this.fields,
    required this.onSave,
    required this.saveLabel,
  });

  final String title;
  final List<Widget> fields;
  final Future<bool> Function() onSave;
  final String saveLabel;

  @override
  State<_FormSheetBody> createState() => _FormSheetBodyState();
}

class _FormSheetBodyState extends State<_FormSheetBody> {
  bool _saving = false;

  Future<void> _save() async {
    setState(() => _saving = true);
    try {
      final ok = await widget.onSave();
      if (ok && mounted) Navigator.of(context).pop(true);
    } finally {
      if (mounted) setState(() => _saving = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final bottomInset = MediaQuery.of(context).viewInsets.bottom;
    return Padding(
      padding: EdgeInsets.fromLTRB(20, 16, 20, 20 + bottomInset),
      child: SingleChildScrollView(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Center(
              child: Container(
                width: 40,
                height: 4,
                margin: const EdgeInsets.only(bottom: 16),
                decoration: BoxDecoration(
                  color: JcTheme.muted.withValues(alpha: 0.4),
                  borderRadius: BorderRadius.circular(2),
                ),
              ),
            ),
            Text(
              widget.title,
              style: const TextStyle(
                  color: JcTheme.text, fontSize: 18, fontWeight: FontWeight.w700),
            ),
            const SizedBox(height: 16),
            ...widget.fields,
            const SizedBox(height: 20),
            Row(
              children: [
                Expanded(
                  child: OutlinedButton(
                    onPressed: _saving ? null : () => Navigator.of(context).pop(false),
                    child: const Text('Cancel'),
                  ),
                ),
                const SizedBox(width: 12),
                Expanded(
                  child: ElevatedButton(
                    onPressed: _saving ? null : _save,
                    child: _saving
                        ? const SizedBox(
                            width: 18,
                            height: 18,
                            child: CircularProgressIndicator(strokeWidth: 2, color: Colors.white),
                          )
                        : Text(widget.saveLabel),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

/// Labelled multiline-capable text field for use inside [showFormSheet].
class FormTextField extends StatelessWidget {
  const FormTextField({
    super.key,
    required this.label,
    required this.controller,
    this.maxLines = 1,
    this.hint,
    this.keyboardType,
  });

  final String label;
  final TextEditingController controller;
  final int maxLines;
  final String? hint;
  final TextInputType? keyboardType;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 14),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(label, style: const TextStyle(color: JcTheme.muted, fontSize: 13)),
          const SizedBox(height: 6),
          TextField(
            controller: controller,
            maxLines: maxLines,
            keyboardType: keyboardType,
            style: const TextStyle(color: JcTheme.text),
            decoration: InputDecoration(hintText: hint, isDense: true),
          ),
        ],
      ),
    );
  }
}

/// Labelled dropdown for use inside [showFormSheet].
class FormDropdown<T> extends StatelessWidget {
  const FormDropdown({
    super.key,
    required this.label,
    required this.value,
    required this.items,
    required this.onChanged,
  });

  final String label;
  final T value;
  final List<DropdownMenuItem<T>> items;
  final ValueChanged<T?> onChanged;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 14),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(label, style: const TextStyle(color: JcTheme.muted, fontSize: 13)),
          const SizedBox(height: 6),
          DropdownButtonFormField<T>(
            initialValue: value,
            isExpanded: true,
            dropdownColor: JcTheme.surfaceAlt,
            items: items,
            onChanged: onChanged,
            decoration: const InputDecoration(isDense: true),
          ),
        ],
      ),
    );
  }
}

/// Labelled on/off switch for use inside [showFormSheet].
class FormToggle extends StatelessWidget {
  const FormToggle({
    super.key,
    required this.label,
    required this.value,
    required this.onChanged,
  });

  final String label;
  final bool value;
  final ValueChanged<bool> onChanged;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 6),
      child: SwitchListTile(
        contentPadding: EdgeInsets.zero,
        title: Text(label, style: const TextStyle(color: JcTheme.text, fontSize: 15)),
        value: value,
        onChanged: onChanged,
        activeThumbColor: JcTheme.accent,
      ),
    );
  }
}

/// Multi-select chip row for use inside [showFormSheet].
class FormChipMulti extends StatelessWidget {
  const FormChipMulti({
    super.key,
    required this.label,
    required this.all,
    required this.selected,
    required this.onChanged,
  });

  final String label;
  final List<String> all;
  final Set<String> selected;
  final ValueChanged<Set<String>> onChanged;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 14),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(label, style: const TextStyle(color: JcTheme.muted, fontSize: 13)),
          const SizedBox(height: 6),
          Wrap(
            spacing: 8,
            runSpacing: 8,
            children: all.map((item) {
              final on = selected.contains(item);
              return FilterChip(
                label: Text(item),
                selected: on,
                onSelected: (v) {
                  final next = Set<String>.from(selected);
                  if (v) {
                    next.add(item);
                  } else {
                    next.remove(item);
                  }
                  onChanged(next);
                },
                backgroundColor: JcTheme.surfaceAlt,
                selectedColor: JcTheme.accent.withValues(alpha: 0.3),
                checkmarkColor: JcTheme.accent,
                labelStyle: TextStyle(color: on ? JcTheme.text : JcTheme.muted),
              );
            }).toList(),
          ),
        ],
      ),
    );
  }
}

File: ios_app/JarvisCopilot/Copilot/UI/FormSheet.swift
import SwiftUI

/// An OPAQUE modal sheet for create/edit forms, ported from `widgets/form_sheet.dart`.
///
/// Opaque (`JcTheme.surface`) is deliberate — a translucent sheet bleeds the aurora
/// backdrop through, the recurring JARVIS-skin gotcha.
///
/// `onSave` returns true on success (the sheet dismisses) or false to stay open
/// (validation failed and the caller surfaced why).
///
/// ```swift
/// .sheet(isPresented: $editing) {
///     FormSheet(title: "New todo", onSave: { await store.save() }) {
///         FormTextField(label: "Title", text: $title)
///     }
/// }
/// ```
struct FormSheet<Fields: View>: View {
    let title: String
    var saveLabel: String = "Save"
    let onSave: () async -> Bool
    @ViewBuilder var fields: Fields

    init(title: String,
         saveLabel: String = "Save",
         onSave: @escaping () async -> Bool,
         @ViewBuilder fields: () -> Fields) {
        self.title = title
        self.saveLabel = saveLabel
        self.onSave = onSave
        self.fields = fields()
    }

    @Environment(\.dismiss) private var dismiss
    @State private var saving = false

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                Text(title)
                    .font(.system(size: 18, weight: .bold))
                    .foregroundStyle(JcTheme.text)
                    .padding(.bottom, 16)
                fields
                HStack(spacing: 12) {
                    Button("Cancel") { dismiss() }
                        .buttonStyle(.bordered)
                        .tint(JcTheme.muted)
                        .frame(maxWidth: .infinity)
                        .disabled(saving)
                    Button {
                        saving = true
                        Task {
                            let ok = await onSave()
                            saving = false
                            if ok { dismiss() }
                        }
                    } label: {
                        if saving {
                            ProgressView().controlSize(.small).tint(.white)
                        } else {
                            Text(saveLabel)
                        }
                    }
                    .buttonStyle(.borderedProminent)
                    .tint(JcTheme.accent)
                    .frame(maxWidth: .infinity)
                    .disabled(saving)
                }
                .padding(.top, 20)
            }
            .padding(20)
        }
        .background(JcTheme.surface)
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
    }
}

/// An uppercase muted field label — the form-field caption.
struct FormFieldLabel: View {
    let text: String
    init(_ text: String) { self.text = text }

    var body: some View {
        Text(text.uppercased())
            .font(.system(size: 11.5, weight: .bold))
            .kerning(0.6)
            .foregroundStyle(JcTheme.muted)
    }
}

/// Labelled, optionally multiline text field for use inside a `FormSheet`.
struct FormTextField: View {
    let label: String
    @Binding var text: String
    var hint: String = ""
    var lineLimit: Int = 1

    var body: some View {
        VStack(alignment: .leading, spacing: 7) {
            FormFieldLabel(label)
            Group {
                if lineLimit > 1 {
                    TextField(hint, text: $text, axis: .vertical).lineLimit(lineLimit...)
                } else {
                    TextField(hint, text: $text)
                }
            }
            .jcFieldStyle()
        }
        .padding(.bottom, 16)
    }
}

/// Labelled select for use inside a `FormSheet` — opens the themed picker sheet.
struct FormDropdown<Value: Hashable>: View {
    let label: String
    @Binding var selection: Value
    let options: [PickerOption<Value>]
    var sheetTitle: String? = nil

    var body: some View {
        VStack(alignment: .leading, spacing: 7) {
            FormFieldLabel(label)
            PickerField(selection: $selection, options: options,
                        hint: label, sheetTitle: sheetTitle ?? label)
        }
        .padding(.bottom, 16)
    }
}

/// Labelled on/off switch for use inside a `FormSheet`.
struct FormToggle: View {
    let label: String
    @Binding var isOn: Bool

    var body: some View {
        Toggle(isOn: $isOn) {
            Text(label).font(JcText.body).foregroundStyle(JcTheme.text)
        }
        .tint(JcTheme.accent)
        .padding(.bottom, 6)
    }
}

/// Multi-select chip row for use inside a `FormSheet`.
struct FormChipMulti: View {
    let label: String
    let all: [String]
    @Binding var selected: Set<String>

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(label).font(.system(size: 13)).foregroundStyle(JcTheme.muted)
            JcWrap(spacing: 8) {
                ForEach(all, id: \.self) { item in
                    let on = selected.contains(item)
                    Button {
                        if on { selected.remove(item) } else { selected.insert(item) }
                    } label: {
                        HStack(spacing: 5) {
                            if on {
                                Image(systemName: "checkmark")
                                    .font(.system(size: 10, weight: .bold))
                                    .foregroundStyle(JcTheme.accent)
                            }
                            Text(item).font(JcText.small)
                        }
                        .foregroundStyle(on ? JcTheme.text : JcTheme.muted)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 7)
                        .background(on ? JcTheme.accent.opacity(0.3) : JcTheme.surfaceAlt, in: Capsule())
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .padding(.bottom, 14)
    }
}

/// Flutter's `Wrap` — lay children out in rows, wrapping at the container width.
/// SwiftUI has no equivalent container, so this is a small `Layout`.
struct JcWrap: Layout {
    var spacing: CGFloat = 8
    var runSpacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let width = proposal.width ?? .infinity
        var rows: CGFloat = 1, x: CGFloat = 0, rowHeight: CGFloat = 0, total: CGFloat = 0
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x > 0, x + spacing + size.width > width {
                total += rowHeight + runSpacing
                rows += 1; x = 0; rowHeight = 0
            }
            x += (x > 0 ? spacing : 0) + size.width
            rowHeight = max(rowHeight, size.height)
        }
        _ = rows
        return CGSize(width: proposal.width ?? x, height: total + rowHeight)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize,
                       subviews: Subviews, cache: inout ()) {
        var x = bounds.minX, y = bounds.minY, rowHeight: CGFloat = 0
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x > bounds.minX, x + spacing + size.width > bounds.maxX {
                y += rowHeight + runSpacing
                x = bounds.minX; rowHeight = 0
            }
            if x > bounds.minX { x += spacing }
            view.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(size))
            x += size.width
            rowHeight = max(rowHeight, size.height)
        }
    }
}

extension View {
    /// The frosted input decoration from `theme.dart`'s `inputDecorationTheme`:
    /// glass fill, 14pt radius, hairline border.
    func jcFieldStyle(radius: CGFloat = JcTheme.fieldRadius) -> some View {
        self
            .font(JcText.body)
            .foregroundStyle(JcTheme.text)
            .textFieldStyle(.plain)
            .padding(.horizontal, 16)
            .padding(.vertical, 14)
            .background(JcTheme.glassFill, in: RoundedRectangle(cornerRadius: radius, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: radius, style: .continuous)
                .strokeBorder(JcTheme.glassBorder, lineWidth: 1))
    }
}
